src/adapters/compound-v2/products/borrow-market/compoundV2BorrowMarketAdapter.ts

```diff
@@ -0,0 +1,355 @@
+  ProtocolTokenUnderlyingRate,
+} from '../../../../types/adapter'
+import { Erc20Metadata } from '../../../../types/erc20Metadata'
+import { IProtocolAdapter } from '../../../../types/IProtocolAdapter'
+import { Protocol } from '../../../protocols'
+import { buildMetadata } from '../../common/buildMetadata'
+import { contractAddresses } from '../../common/contractAddresses'
+import {
+  CUSDCv3__factory,
+  Cerc20__factory,
+  Comptroller__factory,
+} from '../../contracts'
+import { BorrowEvent, RepayBorrowEvent } from '../../contracts/Cerc20'
+
+export class CompoundV2BorrowMarketAdapter implements IProtocolAdapter {
+  chainId: Chain
+  protocolId: Protocol
+  productId: string = 'borrow-market'
+
+  protected provider: CustomJsonRpcProvider
+
+  adaptersController: AdaptersController
+
+  constructor({
+    provider,
+    chainId,
+    protocolId,
+    adaptersController,
+  }: ProtocolAdapterParams) {
+    this.provider = provider
+    this.chainId = chainId
+    this.protocolId = protocolId
+    this.adaptersController = adaptersController
+  }
+
+  getProtocolDetails(): ProtocolDetails {
+    return {
+      protocolId: this.protocolId,
+      name: 'CompoundV2',
+      description: 'CompoundV2 borrow market adapter',
+      siteUrl: 'https:',
+      iconUrl: 'https://',
+      positionType: PositionType.Borrow,
+      chainId: this.chainId,
+      productId: this.productId,
+      assetDetails: {
+        type: AssetType.NonStandardErc20,
+      },
+    }
+  }
+
+  @CacheToFile({ fileKey: 'protocol-token' })
+  async buildMetadata() {
+    return await buildMetadata({
+      chainId: this.chainId,
+      provider: this.provider,
+    })
+  }
+
+  async getProtocolTokens(): Promise<Erc20Metadata[]> {
+    return Object.values(await this.buildMetadata()).map(
+      ({ protocolToken }) => protocolToken,
+    )
+  }
+
+  private async fetchPoolMetadata(protocolTokenAddress: string) {
+    const poolMetadata = (await this.buildMetadata())[protocolTokenAddress]
+
+    if (!poolMetadata) {
+      logger.error(
+        {
+          protocolTokenAddress,
+          protocol: this.protocolId,
+          chainId: this.chainId,
+          product: this.productId,
+        },
+        'Protocol token pool not found',
+      )
+      throw new Error('Protocol token pool not found')
+    }
+
+    return poolMetadata
+  }
+
+  @ResolveUnderlyingPositions
+  async getPositions({
+    userAddress,
+    blockNumber,
+    protocolTokenAddresses,
+  }: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const comptrollerContract = Comptroller__factory.connect(
+      contractAddresses[this.chainId]!.comptrollerAddress,
+      this.provider,
+    )
+
+    const pools = await comptrollerContract.getAssetsIn(userAddress)
+
+    return await filterMapAsync(pools, async (poolContractAddress) => {
+      if (
+        protocolTokenAddresses &&
+        !protocolTokenAddresses.includes(poolContractAddress)
+      ) {
+        return undefined
+      }
```

Comment:
There's not much choice in this case.

It's a filter, so you either leave it as undefined and don't use it, or if it's instantiated as an array you filter using it.

If we set it to default to `[]`, instead of checking for it not to be undefined, we would need to check that the length is not `0`, which is more cumbersome.